service/storage_service.cc

```diff
@@ -6176,6 +6176,162 @@ static bool increases_replicas_per_rack(const locator::topology& topology, const
     return m[dst_rack] + 1 > max;
 }
 
+future<service::group0_guard> storage_service::wait_for_tablet_migration() {
+    auto guard = co_await _group0->client().start_operation(_group0_as, raft_timeout{});
+
+    while (_topology_state_machine._topology.is_busy()) {
+        const auto tstate = *_topology_state_machine._topology.tstate;
+        if (tstate == topology::transition_state::tablet_draining ||
+            tstate == topology::transition_state::tablet_migration) {
+            break;
+        }
+        rtlogger.debug("wait_for_tablet_migration(): topology state machine is busy: {}", tstate);
+        release_guard(std::move(guard));
+        co_await _topology_state_machine.event.wait();
+        guard = co_await _group0->client().start_operation(_group0_as, raft_timeout{});
+    }
+
+    co_return guard;
+}
+
+future<bool> storage_service::exec_tablet_migration(service::group0_guard guard, std::vector<canonical_mutation> updates, sstring reason) {
```

Comment:
The return value means if the updates are added or not. It is up to the caller to decide what to do. It could be stop the loop or other things. I will hold on the change of stop_iteration for now because we might have different ways to issue the request anyway.

Reply:
The only reason they were not added, is because the guard is invalid and operation should be retried. Otherwise, there would be an exception thrown.